src/domain/common/lifecycle/lifecycle.service.ts

```diff
@@ -0,0 +1,125 @@
+import { LogContext } from '@common/enums';
+import {
+  EntityNotFoundException,
+  InvalidStateTransitionException,
+} from '@common/exceptions';
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { State, createMachine, interpret, MachineOptions } from 'xstate';
+import { FindOneOptions, Repository } from 'typeorm';
+import { Lifecycle } from './lifecycle.entity';
+import { ILifecycle } from './lifecycle.interface';
+import { LifecycleEventInput } from './lifecycle.dto.transition';
+
+@Injectable()
+export class LifecycleService {
+  constructor(
+    @InjectRepository(Lifecycle)
+    private lifecycleRepository: Repository<Lifecycle>
+  ) {}
+
+  async createLifecycle(config: string): Promise<ILifecycle> {
+    const lifecycle = new Lifecycle(config);
+    return await this.lifecycleRepository.save(lifecycle);
+  }
+
+  async event(
+    lifecycleEventData: LifecycleEventInput,
+    options: Partial<MachineOptions<any, any>>
+  ): Promise<ILifecycle> {
+    const lifecycle = await this.getLifecycleByIdOrFail(lifecycleEventData.ID);
+    const eventName = lifecycleEventData.eventName;
+    const machineDef = JSON.parse(lifecycle.machine);
+
+    const machine = createMachine(machineDef, options);
+    const machineWithContext = machine.withContext({
+      ...machine.context,
+    });
+    const restoredStateDef = this.getRestoredStateDefinition(lifecycle);
+    const restoredState = State.create(restoredStateDef);
+
+    const nextStates = machineWithContext.resolveState(restoredStateDef)
+      .nextEvents;
+    if (
+      !nextStates.find(name => {
+        return name === eventName;
+      })
+    ) {
+      throw new InvalidStateTransitionException(
+        `Unable to update state: provided event (${eventName}) not in valid set of next events: ${nextStates}`,
+        LogContext.LIFECYCLE
+      );
+    }
+
+    const machineService = interpret(machineWithContext).start(restoredState);
+
+    machineService.send({
+      type: eventName,
+      parentID: machineDef.context.parentID,
```

Comment:
For discussion - are we mandating the parentID?

Reply:
Yes lets discuss